src/rust/ensogl/lib/shape-utils/src/component_color.rs

```diff
@@ -0,0 +1,230 @@
+    fn from(color:color::Color<C>) -> Source {
+        let color = color.into();
+        Source::Static {color}
+    }
+}
+
+impl From<Path> for Source {
+    fn from(path:Path) -> Self {
+        Source::Theme {path}
+    }
+}
+
+
+
+// ===================
+// === Color State ===
+// ===================
+
+/// Indicates which state of the color should be displayed.
+///
+/// Note: Can be extended with new states in the future, for example, "Highlight" or "Achromatic".
+#[derive(Clone,Copy,Debug)]
+pub enum State {
+    /// The default color state.
+    Base,
+    /// A dimmed, muted version of the color. Used to indicate inactivity.
+    Dim,
+    /// Invisible color with alpha set to 0.
+    Transparent,
+}
+
+impl Default for State {
+    fn default() -> Self {
+        State::Base
+    }
+}
+
+
+
+// ===========
+// === Frp ===
+// ===========
+
+ensogl_text::define_endpoints! {
+    Input {
+        source (Source),
+        state  (State),
+    }
+    Output {
+        color  (color::Rgba),
+    }
+}
+
+
+
+// =============
+// === Model ===
+// =============
+
+#[derive(Clone,Debug)]
+struct Model {
+    color_source : RefCell<Option<Source>>,
+    // FIXME[MM] : Replace style watch when #795 is resolved with whatever replaces it.
+    styles       : StyleWatch
+}
+
+impl Model {
+    fn new(app:&Application) -> Self {
+        let color_source = default();
+        let styles       = StyleWatch::new(&app.display.scene().style_sheet);
+        Self{color_source,styles}
+    }
+
+    fn set_source(&self, source:Source) {
+        self.color_source.replace(Some(source));
+    }
+
+    fn get_base_color(&self) -> color::Lcha {
+        match self.color_source.borrow().clone() {
+            Some(Source::Static{color}) => color.into(),
+            Some(Source::Theme{path})   => self.styles.get_color(path),
+            None                        => DEFAULT_COLOR.into()
+        }
+    }
+
+    /// Return the modified version of the base color.
+    fn get_color_dim(&self) -> color::Lcha {
+        match self.color_source.borrow().clone() {
+            Some(Source::Static{color}) => {
+                self.styles.make_color_dim(color).into()
+            },
+            Some(Source::Theme{path})   => {
+                self.styles.get_color_dim(path)
+            },
+            None => DEFAULT_COLOR.into(),
+        }
+    }
+
+    fn get_color_transparent(&self) -> color::Lcha {
+        let mut base    = self.get_base_color();
+        base.data.alpha = 0.0;
+        base
+    }
+}
+
+
+
+// =====================
+// === Dynamic Color ===
+// =====================
+
+/// The `DynamicColor` provides color information through an FRP api. It allows for dynamic color
+/// transitions between different states (e.g, dim or not dim) that are emitted like an animation.
+#[derive(Clone,CloneRef,Debug)]
+pub struct ComponentColor {
+    /// Public FRP api.
+    pub frp   : Frp,
+        model : Rc<Model>,
+}
```

Comment:
I still don't like this abstraction :( It is not as composable as pure frp-based solution, look, now you have such code:

```rust
let edge_color = ComponentColor::new(app);

// This is boilerplate - we should not need such conversions at all.
// These lines are here only to make the types match, they do not add any logic here.
edge_color_source <- input.set_color.map(|&color| {
    component_color::Source::Static{color}
});

// This is boilerplate too. It is used to set color of a color ...
eval edge_color_source ((source) edge_color.frp.source(source.clone()));

// Here we areforced to use some data types, which we need to import.
// I'm not in favor of datatypes if the same effect you can get just by using
// a method.
[...]
    {
        edge_color.frp.state(component_color::State::Dim);
    } else {
        edge_color.frp.state(component_color::State::Base);
    }
[...]

// Here again we are accessing color of a color.
eval edge_color.frp.color ((color) model.set_color(color.into()));
```

Which IMO is overly complex and adds a lot of boilerplate on the usage side, which is bad. Consider instead defining a helper function, like this:

```rust
fn color_animation(network:&frp::Network, source:frp::Stream<color::Lcha>) -> Animation::<color::Lcha> {
    let mut initialized = false;
    let animation       = Animation::<color::Lcha>::new(&network);
    frp::extend! { network
        eval source ([animation](color) {
            if !initialized {
                initialized = true;
                animation.set_value(color);
            } else {
                animation.set_target_value(color);
            }
        })
    }
    animation
}
```

And NOTHING more, no datatypes, nothing. Then, the whole above code would just become the following lines:

```rust
edge_color_dimmed <- edge_color_not_dimmed.map(f!(|color| style.dim(color)));
edge_color        <- switch(&cond,&edge_color_not_dimmed,&edge_color_dimmed);
edge_color_anim   <- color_animation(&network,edge_color);
```

Which is much simpler, much more modular and composable in my opinion. If you feel toherwise, I'd love to deeply discuss this topic tomorrow, maybe im missing something here. Moreover, this design adds few more things:
1. The output type is still `Animation<Lcha>`, which is common, and user does not need to know new abstraction.
2. You can use ANY color transformation here, including your own (in the place of `style.dim(color)`), so no need to define a closed set of transformations.
3. If you want to use style-based color, you should be able to just do in the future something like `style.color("name")`, which should resolve to frp endpoint, so again, no need to create special abstraction for it here.

So, as a result we have much simpler implementation, much less boilerplate on usage side, and a more modular design. I dont see any pros of the proposed solution, but again, I might be missing something.